### PR TITLE
quiche: implement per-event loop processing of buffered QUIC CHLOs

### PR DESCRIPTION
**Pull request by bencebeky**
Description: add UdpListenerCallbacks::onReadReady() and call
quic::QuicDispatcher()::ProcessBufferedChlos() from
ActiveQuicListener::onReadReady() to enable delayed opening of sessions
with a per-event loop limit.

Fixes: #8756

Risk Level: low, not in use

Testing: UdpListenerImplTest is updated to test new onReadReady() calls.
EnvoyQuicDispatcherTest is updated to test that CHLOs are buffered.

Docs Changes: n/a
Release Notes: n/a

Signed-off-by: Bence Béky <bnc@google.com>

### CODE REVIEW

File: test/extensions/quic_listeners/quiche/envoy_quic_dispatcher_test.cc [hunk 1]
```diff
@@ -177,20 +189,107 @@ TEST_P(EnvoyQuicDispatcherTest, CreateNewConnectionUponCHLO) {
...
+      // Stop iteration to avoid calling getRead/WriteBuffer().
+      .WillOnce(Return(Network::FilterStatus::StopIteration));
+  EXPECT_CALL(network_connection_callbacks, onEvent(Network::ConnectionEvent::Connected));
+
+  quic::QuicBufferedPacketStore* buffered_packets =
+      quic::test::QuicDispatcherPeer::GetBufferedPackets(&envoy_quic_dispatcher_);
+  EXPECT_FALSE(buffered_packets->HasChlosBuffered());
+  EXPECT_FALSE(buffered_packets->HasBufferedPackets(connection_id_));
+
+  // Incoming CHLO packet is buffered.
```
Comment (danzh2010) [inline test/extensions/quic_listeners/quiche/envoy_quic_dispatcher_test.cc]: nits: document why the first CHLO is buffered immediately.

File: test/extensions/quic_listeners/quiche/active_quic_listener_test.cc [hunk 2]
```diff
@@ -44,39 +47,136 @@ class ActiveQuicListenerPeer {
...
+      // calls for the same object.
+      EXPECT_CALL(filter_chain_manager_, findFilterChain(_))
+          .InSequence(seq)
+          .WillOnce(Return(&filter_chains_.back()));
+    }
+  }
+
+  // TODO(bencebeky): Factor out parts common with
+  // EnvoyQuicDispatcherTest::createFullChloPacket() to test_utils.
+  void GenerateCHLO(quic::QuicConnectionId connection_id) {
```
[the remark this answers is not in the document]
Comment (bencebeky) [reply test/extensions/quic_listeners/quiche/active_quic_listener_test.cc]: Done.

File: source/extensions/quic_listeners/quiche/active_quic_listener.h [hunk 3]
```diff
@@ -19,6 +19,9 @@ class ActiveQuicListener : public Network::UdpListenerCallbacks,
                            public Server::ConnectionHandlerImpl::ActiveListenerImplBase,
                            Logger::Loggable<Logger::Id::quic> {
 public:
+  // TODO(bencebeky): Tune this value.
+  static const size_t kNumSessionsToCreatePerEpoll = 16;
```
[the remark this answers is not in the document]
Comment (bencebeky) [reply source/extensions/quic_listeners/quiche/active_quic_listener.h]: Done.

File: test/extensions/quic_listeners/quiche/envoy_quic_dispatcher_test.cc (hunk 1 above)
Comment (bencebeky) [reply test/extensions/quic_listeners/quiche/envoy_quic_dispatcher_test.cc]: Done.

**Comment by bencebeky** [comment]
> Yeah! LGTM overall. Feel free to address the nits in a follow up PR! Thanks for the major refactoring in tests.

Thank you for the review.

**Comment by mattklein123** [comment]
Can you check the GCC build?

/wait

**Comment by mattklein123** [comment]
Do you mind merging master to hopefully fix the OSX build?

/wait

**Comment by bencebeky** [comment]
> Do you mind merging master to hopefully fix the OSX build?
> 
> /wait

Thank you for your advice.  Done, and indeed this fixed the OSX build.

**Pull request merged**